embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceApiService.kt

```diff
@@ -170,38 +172,45 @@ internal class EmbraceApiService(
         return postOnExecutor(bytes, request, onComplete)
     }
 
+    /**
+     * Posts an API call to the API.
+     * If the device is offline or the endpoint is rate limited, the API call is saved to be sent later.
+     * If the API call fails, it is retried and if it succeeds, it is removed from cache.
+     */
     private fun postOnExecutor(
         payload: ByteArray,
         request: ApiRequest,
         onComplete: (() -> Any)?
     ): Future<*> {
         return executorService.submit(
             NetworkRequestRunnable(request) {
-                try {
-                    if (lastNetworkStatus != NetworkStatus.NOT_REACHABLE) {
-                        executePost(request, payload)
+                val pendingApiCall = pendingApiCallsSender.savePendingApiCall(request, payload)
+
+                if (lastNetworkStatus != NetworkStatus.NOT_REACHABLE &&
+                    !rateLimitHandler.isRateLimited(request.url.endpoint())
+                ) {
+                    // Execute the request if the device is online and the endpoint is not rate limited.
+                    val response: ApiResponse = executePost(request, payload)
+
+                    if (response is ApiResponse.Success) {
+                        // Remove the API call from cache if it was sent successfully.
+                        pendingApiCallsSender.removePendingApiCall(pendingApiCall)
+                        onComplete?.invoke()
                     } else {
-                        pendingApiCallsSender.scheduleApiCall(request, payload)
-                        logger.logWarning("No connection available. Request was queued to retry later.")
+                        // Schedule the API call to be sent later if it should be retried.
+                        if (pendingApiCallsSender.shouldRetry(response)) {
```

Comment:
What if there are no retry jobs outstanding to check for requests to retry? If `pendingApiCallsSender.scheduleApiCall()` isn't called, it won't check for that case and schedule a job.